src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            await buffer.Ensure(sizeof(int), async);
+
+            var fieldCount = buffer.ReadInt32();
+            if (fieldCount != _members.Length)
+                throw new InvalidOperationException($"pg_attributes contains {_members.Length} fields for type {PgDisplayName}, but {fieldCount} fields were received.");
+
+            var composite = new T();
+            if (ByReference<T>.IsTrue)
+            {
+                var reference = new ByReference<T>(composite);
+                foreach (var member in _members)
+                    await member.Read(reference, buffer, async);
+
+                return reference.Value;
+            }
+            else
+            {
+                foreach (var member in _members)
+                    await member.Read(composite, buffer, async);
+
+                return composite;
+            }
+        }
+
+        public override async Task Write(T value, NpgsqlWriteBuffer buffer, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
+        {
+            if (buffer.WriteSpaceLeft < sizeof(int))
+                await buffer.Flush(async);
+
+            buffer.WriteInt32(_members.Length);
+
+            foreach (var member in _members)
+                await member.Write(value, buffer, lengthCache, async);
+        }
+
+        public override int ValidateAndGetLength(T value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
+        {
+            if (lengthCache == null)
+                lengthCache = new NpgsqlLengthCache(1);
+
+            if (lengthCache.IsPopulated)
+                return lengthCache.Get();
+
+            // Leave empty slot for the entire composite type, and go ahead an populate the element slots
+            var position = lengthCache.Position;
+            lengthCache.Set(0);
+
+            // number of fields + (type oid + field length) * member count
+            var length = sizeof(int) + sizeof(int) * 2 * _members.Length;
+            foreach (var member in _members)
+                length += member.ValidateAndGetLength(value, ref lengthCache);
+
+            return lengthCache.Lengths[position] = length;
+        }
+
+        public static MappedCompositeHandler<T> Create(PostgresCompositeType postgresType, ConnectorTypeMapper typeMapper, INpgsqlNameTranslator nameTranslator)
+        {
+            var compositeType = typeof(T);
+            var mappedFields = postgresType.Fields;
+            var mappedMembers = new CompositeMemberInfo<T>[mappedFields.Count];
+            var mappedMemberCount = 0;
+
+            var memberInfoType = ByReference<T>.IsTrue
+                ? typeof(CompositeMemberInfoOfStruct<,>)
+                : typeof(CompositeMemberInfoOfClass<,>);
+
+            foreach (var compositeMember in compositeType.GetMembers(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
```

Comment:
I have the same concerns about that too, but for compatibility did it this way. If you and @austindrenski are okay with dropping non-public member support, will do it.

Reply:
I lean toward not including non-public members by default. It seems both simpler and more appropriate to just respect the CLR visibility, barring explicit info from the user that "yes, I _do_ want to map this specific non-public member".

If we want to support non-public members, maybe we can look into (and try to match) the EF Core decisions on non-public mapping?

Reply:
Okay, will drop in a new PR to track that in issues and release notes. Otherwise, there is a change that we will forget about that and won't notice our users. Opened #2405 to track progress.

Reply:
I took a look to confirm and [we currently look at public members only](https://github.com/npgsql/npgsql/blob/94b16acd54c805b74226a002a4f2f3816020b05f/src/Npgsql/TypeHandlers/UnmappedCompositeHandler.cs#L224). This wouldn't be a breaking change.

Reply:
Yes, fixed a test.